### ISSUE TITLE
Use variable size integer in the OSBTreeBonsai buckets

### ISSUE THREAD

**Issue by tglman** [orientechnologies/orientdb#4500]
The today implementation of a OSBTreeBonsai bucket it's separated in two parts

an header consisted of an ordered array of pointers at the top of the bucket where each pointer is an int(4 bytes)

an flat list of values at the bottom of the bucket that consist in two type of content depending on the type of bucket:
- node bucket
  a node pointer consisted by a long(8 bytes)+int(4 bytes)
- leaf bucket
  a key that is a rid short(2 bytes) + long(8 bytes) plus a value int(4 bytes)

we could use variable size integer for the bottom of the bucket, that can save up to 30% of space for int and 50% for long.
so we could store almost a 40% more entry in each bucket.

another optimization could be done on the top of the bucket, we could force the maximum limit of size of each bucket to 64kb and in that case we cold save/use 50% of more space, the today bucket size is 2kb.

_binary compatibility_
value optimization
- node bucket
  the code could be refactored to make use of a serializer also in case of node bucket and drop back to a backward compatible when the serializer is not found(i've to double check the code thought).
- leaf bucket
  already use the serializer, we can drop in a variable size serializer and refactor a bit the code for support variable size

header optimization
 can be done just using a complete new format(or just a new file extension), as far as i know there is no version management for the bucket structure

be aware that variable size integer use much more cpu compared to drop in integers.

**Comment by laa** [comment]
About , var int, it may be reasonable sure. About "another optimization could be done on the top of the bucket, we could force the maximum limit of size of each bucket to 64kb and in that case we cold save/use 50% of more space, the today bucket size is 2kb". It was done in first version of implementation and as result performance was much worse. I explain why. You see, rids have very small size and as result ridbags do not fill whole page, only part of this. So density of stored information 
is very low as result we wasted  a lot of IO operations and it was reflected in benchmarks. As result we decided to improve density of information stored in one bonsaitree page by using mini pages, and numbers difference became much better.

**Comment by laa** [comment]
@tglman About var int actually reasoning was the same as you provided. I lets look on this results for example http://mechanical-sympathy.blogspot.com/2012/07/native-cc-like-performance-for-java.html but I think it worth a shot. ))

**Comment by laa** [comment]
Actually I think if we will use var int for pages, we should not increase but decrease size of mini page. To provide the same density of information for one page.

**Comment by tglman** [comment]
@laa i express myself wrongly about the header, by idea is to bring the header from be int[] to be short[], doing that the page cannot go over a size of 64kb the today allowed size is 2gb, so with that we could keep everything even smaller

**Comment by laa** [comment]
I do not talk about header, I mean, correct me if I wrong, that you  propose to make bucket size 64 KB instead of 2 KB is ti correct ?

**Comment by tglman** [comment]
nope i propose to not change much the today bucket size setting(or anyway not doing it without proper performace check), but change the size of the header pointers inside, because if the bucket it's smaller we can keep smaller pointers and save more data in the bucket itself.

**Comment by laa** [comment]
that is reasonable. But any way I propose to try to make bucket size even smaller after that. It may increase performance.

**Comment by tglman** [comment]
@laa  do you have any idea how to do this changes in backward compatible way ? do you think it make sense create a new file extension for a new ridbag so at the moment of read we know how to read ? any better idea ?

**Comment by laa** [comment]
Yes, we can use the same approach which we used for hash index, we have 2-nd version of hash index with different data structures but it does not affect binary compatibility.
It was done in following way:

1. We have created new class com.orientechnologies.orient.core.index.hashindex.local.OLocalHashTable20 and copied old implementation there.
2. We created new implementation of com.orientechnologies.orient.core.index.hashindex.local.OLocalHashTable .
3. Together with reference to data structure we store version of data structure. If version is absent or less than 2 we load index data using OLocalHashTable20, otherwise we use OLocalHashTable .

It made data structure changes completely transparent to users and allowed to support old data structures without much of overhead. I mean if we change disk cache we do not analyze complex if branches but merely change each data structure as independent one.
